src/serialize.h

```diff
@@ -148,7 +148,14 @@ enum
     SER_GETHASH         = (1 << 2),
 };
 
-#define READWRITE(...)      (::SerReadWriteMany(s, ser_action, __VA_ARGS__))
+//! Convert the reference base type to X, without changing constness or reference type.
+template<typename X> X& ReadWriteAsHelper(X& x) { return x; }
+template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
+template<typename X> X&& ReadWriteAsHelper(X&& x) { return std::move(x); }
```

Comment:
I think this may be broken in the case where `x` binds to the temporary result of an implicit conversion, e.g. in deserialization of `ServiceFlags nServices` above as `int64_t`, or in a more straightforward case of a `char` variable being deserialized as an int. In these cases the deserialization will just happen into a temporary variable, and the actual `obj` argument passed to `READWRITEAS` won't get updated.

This could be fixed with a serialization wrapper class that can deserialize into a temporary and then save the results on destruction. But a simpler alternative might just be to drop the two rvalue reference overloads of `ReadWriteAsHelper` so it would be an error to call `READWRITEAS` anywhere where deserialization would need to happen into a temporary variable. It seems like the only place where this happens is with `nServices`